docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
 
-The eBPF for Windows project is designed to permit anyone to add new hooks, program types or helper functions,
-without the need to modify either the eBPF execution context or the eBPF verifier.  We use the term "eBPF extension"
-to mean a driver or component that exposes new eBPF hooks or helper functions.
+## 1.1 Windows Network Module Registrar
 The eBPF for Windows project uses the
-[Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-module-registrar2)
-to decouple the eBPF extensions from the core eBPF for Windows framework.  In the NMR architecture,
+[Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar)
+feature in Windows to develop eBPF extensions that are decoupled from the core eBPF for Windows framework.  NMR facilitates software
+modules to be attached to each other. The
 [Network Programming Interface (NPI)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-programming-interface)
-contracts are identified by a GUID.
+defines the contract between the modules.  The eBPF extensions would implement the
+[Provider Modules](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/provider-module) for two types of NPIs as discussed in detail below.
+Even though the NMR component and the relevant APIs have the word "Network" in them, these software modules
+are completely agnostic of Windows Networking stack and non-networking eBPF extensions can be developed using these.
 
-## NPI Contract for eBPF Program Information
+## 1.2 Prerequisites
+The audience is assumed to be familiar with [NMR Architecture](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/architecture-overview) and
+the various aspects of developing NMR modules as documented [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-module-registrar2).
 
+## 1.3 NPI Contracts for eBPF Extensions
+eBPF Extensions need to implement *provider modules* for two types of NPIs. They are the **Program Information NPI** provider and the **Hook NPI** provider.
+
+### 1.3.1 eBPF Program Information NPI Provider
 The eBPF program information NPI contract is used to provide information about an eBPF program type. Program types
 are defined as the ABI contracts that eBPF programs are written to conform to.
 This information is consumed by the eBPF verifier to ensure that any eBPF programs of a given type are safe to load
 and execute. The ABI contract includes both a description of the &quot;context&quot; parameter passed to the eBPF
-program as well as the list of specific helper functions that are available to such eBPF programs. eBPF extensions
-should not unload until the last NMR client detaches, signifying that no eBPF programs are using helper functions
-it provides.
+program as well as the list of specific helper functions that are available to such eBPF programs.
```

Comment:
So this means that an extension cannot add a new program type specific helper function and make it accessible to another extension that implements the program information NPI Provider, right?  All program type specific helper functions have to be implemented in the same provider, even if there are multiple hook NPI Provider implementations?

Reply:
This is technically possible. Since the program info NPI provider for both the extensions will use the same NPI ID == Program Type. When EC registers a NPI client with the same NPI ID, NMR will attach both the providers to it. So the EC can aggregate helper functions from both the providers and pass them to verifier/JIT. But that is not implemented today. And I think that will be over-engineering as well. I think we should mandate, there be a single NPI provider for a given program type. Note that it may refer to helper functions that are present on another driver - this is discussed as part of another comment below.

Reply:
I will start a discussion on this. And later convert to issue as needed.